### PR TITLE
RFC for Fn::Length

### PR DESCRIPTION
**Pull request by juegong2**
### Language Enhancement Request For Comment

This is a request for comments about new intrinsic function Fn::Size. See https://github.com/aws-cloudformation/cfn-language-discussion/issues/70 for additional details. 

---

## Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: RFCs/0070-Fn::Size.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# RFC for Fn::Size
+
+* **Original Author(s):**: @juegong2
+* **Tracking Issue**: https://github.com/aws-cloudformation/cfn-language-discussion/issues/70
+
+# Summary
+
+We will support an intrinsic function called Fn::Size that will return the number of elements in a given list.
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:8]:
```suggestion
We will support an intrinsic function called `Fn::Size` that will return the number of elements in a given list.
```

File: RFCs/0070-Fn::Size.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+# Motivation
+
+CloudFormation users may want to get the number of elements in a given list to set certain resource parameters or generate conditions. It may also help to check if index is out of bound for !Select function. This RFC proposes to support the use case of getting the number of elements of a given list.
+
+* Reference:
+  * https://github.com/aws-cloudformation/cfn-language-discussion/issues/61
+  * https://github.com/aws-cloudformation/cfn-language-discussion/issues/37
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:16]: Reading through this issue, I'm unclear how `Fn::Size` solves the customer problem described. From the comments, it seems like there was consensus that `Fn::For` would be a cleaner solution to the original problem. Do you have an example of how `Fn::Size` would be used to solve the problem described in this specific issue?

File: RFCs/0070-Fn::Size.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        3
+      ]
+    }
+  }
+}
+```
+
+# Limitation
+
+`Fn::Size` does not support reference to resource properties. It only supports reference to Parameters.
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:71]: I'm a little confused by this statement. Are there any resource types with return values of type list? I'm pretty sure they're all strings... Should be more precise and say `Fn::Size` will not work with lists that are dynamically generated at provisioning time. For example, assuming MyBucket is the logical id of a `AWS::S3::Bucket` resource, the following is not supported:

```yaml
Fn::Size:
  Fn::Split: [':', !GetAtt MyBucket.Arn]
```

because the ARN of MyBucket is not known until during provisioning.

File: RFCs/0070-Fn::Size.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+# Details
+
+`Fn::Size` is an intrinsic function that takes in a list and returns the size of the given list. The list could be:
+
+* hardcoded list in template
+* array generated by intrinsic functions like `Fn::Split`
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:78]:
```suggestion
* array generated by intrinsic functions like `Fn::Split` of a hardcoded string or of a `Ref` to a string parameter
```

Need to be careful here. The initial version won't work in all cases of `Fn::Split`. See my previous comment.

File: RFCs/0070-Fn::Size.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+* `Ref` to a Parameter of type `CommaDelimitedList`
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:79]:
```suggestion
* `Ref` to a parameter of type `CommaDelimitedList`
```

File: RFCs/0070-Fn::Size.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+* `Ref` to a Parameter of type `List<Number>`
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:80]:
```suggestion
* `Ref` to a parameter of type `List<Number>`
```

File: RFCs/0070-Fn::Size.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+* `Ref` to a Parameter of type `List<AWS-specific parameter types>`
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:81]:
```suggestion
* `Ref` to a parameter of type `List<AWS-specific parameter types>`
```

File: RFCs/0070-Fn::Size.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+* `Ref` to an SSM Parameter of type
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:82]:
```suggestion
* `Ref` to an parameter of type
```

File: RFCs/0070-Fn::Size.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  * `AWS::SSM::Parameter::Value<List<String>>`
+  * `AWS::SSM::Parameter::Value<CommaDelimitedList>`
+  * `AWS::SSM::Parameter::Value<List<AWS-specific parameter type>>`
+
+Within `Fn::Size`, following intrinsic functions will be supported:
+* Condition functions
+* Fn::FindInMap
+* Fn::Base64
+* Fn::Join
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:91]: This has similar caveats to `Fn::Split`

File: RFCs/0070-Fn::Size.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+* Fn::Select
+* Fn::Split
+* Fn::Sub
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:94]: Also similar caveats to `Fn::Split`

**Review by jlhood — CHANGES_REQUESTED** [review]
Great first draft. No issues with syntax. Just want to make sure we include some clarifications on limitations.

File: RFCs/0070-Fn::Size.md (hunk 9 above)
Comment (juegong2) [reply RFCs/0070-Fn::Size.md:91]: Will add constraints to line 87

File: RFCs/0070-Fn::Size.md (hunk 2 above)
Comment (juegong2) [reply RFCs/0070-Fn::Size.md:16]: No I don't. But Fn::Size is suggested as a solution in this issue. Therefore I put the issue here.

File: RFCs/0070-Fn::Size.md (hunk 9 above)
Comment (juegong2) [reply RFCs/0070-Fn::Size.md:96]: updated

File: RFCs/0070-Fn::Size.md (hunk 10 above)
Comment (juegong2) [reply RFCs/0070-Fn::Size.md:99]: updated

File: RFCs/0070-Fn::Size.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+}
+```
+because the ARN of MyBucket is not known until during provisioning.
+
+# Details
+
+`Fn::Size` is an intrinsic function that takes in a list and returns the size of the given list. The list could be:
+
+* hardcoded list in template
+* array generated by intrinsic functions like `Fn::Split` of hardcoded parameters or of `Ref` to parameters
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:83]:
```suggestion
* array generated by intrinsic functions like `Fn::Split` of hardcoded values or of `Ref` to parameters
```

File: RFCs/0070-Fn::Size.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+          "Fn::Size": [ "m5.large", "m5.xlarge" ]
+        },
+        3
+      ]
+    }
+  }
+}
+```
+
+# Limitation
```
Comment (jlhood) [inline RFCs/0070-Fn::Size.md:68]: Thanks for making the wording more precise here. I suspect we'll get feedback on this as from a customer point of view, this is not ideal (ideally should support this case as well). Anticipating that, maybe we should add something like this to the bottom of this section: "Note: This is a short-term limitation due to underlying implementation constraints. In the fullness of time, this limitation should be removed."

**Review by jlhood — CHANGES_REQUESTED** [review]
2 minor comments, then it should be good to go!

File: RFCs/0070-Fn::Size.md (hunk 11 above)
Comment (juegong2) [reply RFCs/0070-Fn::Size.md:83]: here I meant parameter of an intrinsic function, not parameter of the template. But yeah it will cause confusion :(

**Review by jlhood — APPROVED** [review]
Looks good. Thanks for iterating!

**Comment by lorengordon** [comment]
I'd suggest `Fn::Length` for greater similarity with most languages I've seen. I had no idea what was meant by `Size` when I first saw it here. I thought maybe it was related to something like the "size" of an EC2 instance type.

File: README.md [hunk 13]
```diff
@@ -11,6 +11,7 @@ Go to: [What is an RFC](#what-is-an-rfc) | [RFC Process](#rfc-process)
 [9](https://github.com/aws-cloudformation/cfn-language-discussion/issues/9)|[Adding looping functionality in CFN Template](https://github.com/aws-cloudformation/cfn-language-discussion/pull/19)|[@MalikAtalla-AWS](https://github.com/MalikAtalla-AWS)|✍️ review
 [14](https://github.com/aws-cloudformation/cfn-language-discussion/issues/14)|[Convert Template Block to JSON String](https://github.com/aws-cloudformation/cfn-language-discussion/pull/15)|[@mluk-aws](https://github.com/mluk-aws)|✍️ review
 [11](https://github.com/aws-cloudformation/cfn-language-discussion/issues/11)|[Allow Intrinsic Functions and Pseudo-Parameter References in DeletionPolicy and UpdateReplacePolicy](https://github.com/aws-cloudformation/cfn-language-discussion/pull/21)|[@MalikAtalla-AWS](https://github.com/MalikAtalla-AWS)|✍️ review
+[70](https://github.com/aws-cloudformation/cfn-language-discussion/issues/70)|[Adding Fn::Size support in CFN Template](https://github.com/aws-cloudformation/cfn-language-discussion/pull/72)|[@juegong2](https://github.com/juegong2)|✍️ review
```
Comment (jlhood) [inline README.md:14]: Can you put this change in a separate PR? If you couple it with this one it won't get added to the README until after it's fully approved which doesn't make a lot of sense.

**Comment by josb** [comment]
@lorengordon I'm fine with the identifier `Fn::Length` if that is preferred.

**Comment by juegong2** [comment]
Updated the naming of the new intrinsic function to Fn::Length

**Review by jlhood — CHANGES_REQUESTED** [review]
Can you rename the RFC file as well?

**Comment by juegong2** [comment]
File name is updated as well

**Review by jlhood — APPROVED** [review]

**Comment by benbridts** [comment]
I'm not sure if I understand the use case of this function:

- For conditionals it's very clunky without support for integer comparisons.
- For OutOfBound checking it has similar problems:
   - You can't use conditionals (see above)
   - CloudFormation will throw an error either way

File: RFCs/0070-Fn::Length.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+`Fn::Length` will not work with lists that are dynamically generated at provisinoing time. For example, assuming MyBucket is the logical id of a AWS::S3::Bucket resource, the following is not supported:
+```json
+{
+  "Fn::Length": { "Fn::Split": [":", { "Fn::GetAtt": [ "MyBucket", "Arn" ] } ] }
+}
+```
+because the ARN of MyBucket is not known until during provisioning.
+
+**Note:** This is a short-term limitation due to underlying implementation constraints. In the fullness of time, this limitation should be removed.
```
Comment (benbridts) [inline RFCs/0070-Fn::Length.md:78]: I would remove this note, or be very careful about wording. Promising a short-term solution and not meeting this self-imposed deadline is a good way to lose trust.

Additionally, getting  this to be useful during provisioning looks complex from the outside:

Having !Length work during provisioning does seem  similar to how other functions work (so short term might be okay here)
However, most use cases will expect to be able to use an !If with !Length, and this would be a major change in how CloudFormation works. Currently in the If documentation, there is this remark.

> You can only reference other conditions and values from the Parameters and Mappings sections of a template. For example, you can reference a value from an input parameter, but you can't reference the logical ID of a resource in a condition.

Repeating that in the limitations here might be a good idea too

File: RFCs/0070-Fn::Length.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+        3
+      ]
+    }
+  }
+}
+```
+
+# Limitation
+
+`Fn::Length` will not work with lists that are dynamically generated at provisinoing time. For example, assuming MyBucket is the logical id of a AWS::S3::Bucket resource, the following is not supported:
```
Comment (MalikAtalla-AWS) [inline RFCs/0070-Fn::Length.md:70]:
```suggestion
`Fn::Length` will not work with lists that are dynamically generated at provisioning time. For example, assuming MyBucket is the logical id of a AWS::S3::Bucket resource, the following is not supported:
```

File: RFCs/0070-Fn::Length.md (hunk 14 above)
Comment (juegong2) [reply RFCs/0070-Fn::Length.md]: Thanks for your feedback! I have updated the wording of this note.

**Review by MalikAtalla-AWS — APPROVED** [review]

**Review by jlhood — APPROVED** [review]

File: RFCs/0070-Fn::Length.md (hunk 14 above)
Comment (benbridts) [reply RFCs/0070-Fn::Length.md]: 👍🏻

File: RFCs/0070-Fn::Length.md [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+```json
+{
+  "Parameters": {
+    "InstanceTypes": {
+      "Type": "CommaDelimitedList"
+    },
+    "NumberList": {
+      "Type": "List<Number>"
+    }
+  }
```
Comment (MalikAtalla-AWS) [inline RFCs/0070-Fn::Length.md:30]:
```suggestion
  },
```

Comment (MalikAtalla-AWS) [inline RFCs/0070-Fn::Length.md:30]:
```suggestion
  },
```

**Review by jlhood — APPROVED** [review]

**Review by lejiati — APPROVED** [review]
